web: Resolve URLs without a protocol in the extension player and improve error messages

Korne127 opened the pull request:
Currently, only URLs explicitly including the web protocol (HTTPS or HTTP) in the beginning can be resolved in the extension player. Additionally, if the URL doesn't include the HTTPS or HTTP protocol in the beginning, a wrong error is displayed. It states that the user ran Ruffle on the file: protocol which is wrong and confusing:
<img width="1582" alt="Bildschirmfoto 2024-06-28 um 15 53 04" src="https://github.com/ruffle-rs/ruffle/assets/33464213/b01c70c4-d1bf-4a49-8fc6-e551e0368edb">

This pull request fixes this.
With this pull request, the extension player resolves URLs without a protocol. It adds https:// or http:// if the URL is likely a valid web URL and file:/// if the URL is a valid file URL.
To determine whether HTTPS or HTTP should be used, Ruffle tests whether the server supports HTTPS. HTTP is only used if the server only supports HTTP and not HTTPS. This test happens while the loading screen is displayed.

This means that the user can now enter URLs without a leading https:// like you (usually) do in the browser.

Additionally, this pull request fixes and improves the error behaviour if the URL can't be resolved or used.
Instead of the wrong error, one of four more specific errors will be displayed:
- One new error if the given URL is a local file URL:
  <img width="1582" alt="Bildschirmfoto 2024-06-28 um 15 56 26" src="https://github.com/ruffle-rs/ruffle/assets/33464213/fb0686af-52ee-45f0-abb3-79294ef061d6">
- One new error if the given URL uses a different protocol than https, http or file (e.g. ftp):
  <img width="1582" alt="Bildschirmfoto 2024-06-28 um 15 57 44" src="https://github.com/ruffle-rs/ruffle/assets/33464213/f7ddc81d-87ef-4e1b-87e9-08cf6cf71e8a">
- One new error if the given URL can't be resolved to any valid URL for any protocol
  <img width="1582" alt="Bildschirmfoto 2024-06-28 um 15 58 07" src="https://github.com/ruffle-rs/ruffle/assets/33464213/f7db5285-8101-49c9-a278-ca9ece2b53b9">
- And the usual SWF file not found error if the file under the given web URL couldn't be found.
  That error has also been modified to also include the tip that the HTTP protocol should be (explicitly) used for websites that don't support HTTPS:
  <img width="1582" alt="Bildschirmfoto 2024-06-28 um 15 54 12" src="https://github.com/ruffle-rs/ruffle/assets/33464213/bef18bbf-365c-45c4-9486-bda7b619549d">

Also, a bug that caused the document title to not be set correctly if the SWF URL was the main page of a domain has been fixed. The handling of setting the document title of the extension player has generally been made more consistent.

Korne127 committed:
web: Resolve URLs without a protocol in the extension player

Previously, only URLs explicitly including the web protocol (HTTPS or
HTTP) in the beginning could be resolved in the extension player.
This has been changed. The extension player now resolves URLs without a
protocol and adds https:// or http:// if the URL is likely a valid web
URL and file:/// if the URL is a valid file URL.
HTTP is only used if the server only supports HTTP and not HTTPS.

Croworbit commented:
hello Korne, 
useful PR! :)
sorry but there is a typo on the 4th image!

Korne127 commented:
Hey :)
Thanks for the nice response and the feedback :D 
I'm sorry, but I genuinely don't see the typo… can you point it out for me? 😅

danielhjacobs commented on web/packages/core/texts/en-US/messages.ftl:
```diff
@@ -42,9 +42,12 @@ error-wasm-mime-type =
     Ruffle cannot parse the requested file.
     The most likely reason is that the requested file is not a valid SWF.
-error-swf-fetch =
+error-swf-fetch-v2 =
```
Why the addition of the `-v2`? Crowdin will throw out all the old translations regardless, and if you're not also keeping `error-swf-fetch` as a separate translation string I don't see a reason to change this key.

Korne127 replied:
> Crowdin will throw out all the old translations regardless

Huh? I didn't know that. That's actually not really good because this is utilising [a core Fluent feature](https://github.com/projectfluent/fluent/wiki/Fluent-vs-gettext#translation-invalidation): The separation of minor changes that don't affect translations and major changes that do affect them:
> Fluent's use of unique identifiers allows for at least two of those three states to remain separated - minor changes may be applied and if the developer does not alter the unique identifier (a.k.a does not change the social contract), all translations remain valid. On the other hand if the developer changes the ID, all translations become invalid and must be updated.

I wish Fluent would have implemented the idea of message versioning [they're investigating](https://github.com/projectfluent/fluent/wiki/Fluent-vs-gettext#translation-invalidation) as a third option because this is the ideal case in which this would be used (as old translations are still better than none, but should be marked as outdated and updated eventually). However, the [related issue](https://github.com/projectfluent/fluent/issues/141) hasn't been updated in years :/ I might try to get it more attention again in the future.

Korne127 replied:
As Crowdin doesn't follow this Fluent feature (which itself is not good imo), I removed the `-v2`.

Croworbit commented:
> Hey :) Thanks for the nice response and the feedback :D I'm sorry, but I genuinely don't see the typo… can you point it out for me? 😅

"The url blabla test is no valid url"
I expected "the url blablatest is not a valid url"

Korne127 commented:
Hmm, I would have thought that both is valid tbh, although the second one is probably better.
I can change it, but when I went back to look at the messages, I don't really like the repetition. I think I might change it to "Bla-Bla-Test is not a valid URL".
(And change enter to include on the last screenshot.)

Croworbit commented:
amittedly its a nitpick 😅

Korne127 commented:
@Croworbit Don't worry, I really appreciate it! Especially the things that I overlooked myself are good to know :)

evilpie commented on web/packages/extension/src/player.ts:
```diff
@@ -236,22 +239,120 @@ window.addEventListener("load", () => {
+        options: RequestInit,
+        timeout: number,
+    ) {
+        // TODO: Use AbortSignal in the future when it doesn't break browser compatibility
```
Is it really too new?

Korne127 replied:
Yes, it breaks our current [browser compatibility](https://github.com/ruffle-rs/ruffle/wiki)

evilpie replied:
I think we could do `await Promise.race([fetch(), new Promise((_, reject) => setTimeout(reject, timeout))])` maybe?

evilpie replied:
Actually, shouldn't we support it?
https://developer.mozilla.org/en-US/docs/Web/API/Request/signal says
Chrome 66
Firefox 57
Safari 12.1

Or are you hoping to use `AbortSignal.timeout` directly?

Korne127 replied:
I meant something like `fetch(url, { signal: AbortSignal.timeout(…) })`. But its [minimum supported version](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal/timeout_static) is over ours.
But you are right, [AbortController()](https://developer.mozilla.org/en-US/docs/Web/API/AbortController/AbortController), [AbortController#abort()](https://developer.mozilla.org/en-US/docs/Web/API/AbortController/abort) and [Request#signal](https://developer.mozilla.org/en-US/docs/Web/API/Request/signal) are all available for our minimum supported versions, so we could use a polyfill like
```
AbortSignal.timeout ??= function timeout(ms) {
  const ctrl = new AbortController()
  setTimeout(() => ctrl.abort(), ms)
  return ctrl.signal
}
```
([Reference](https://stackoverflow.com/questions/46946380/fetch-api-request-timeout/57888548))
I'll change that :)

evilpie commented on web/packages/extension/src/player.ts:
```diff
@@ -236,22 +239,120 @@ window.addEventListener("load", () => {
+        }
+
+        try {
+            // Only use http if it works but https doesn't
```
I am not totally convinced that we shouldn't just default to https:

lvyitian replied:
That's because HTTP website cannot fetch HTTPS resources, and vice versa.

evilpie replied:
How is this relevant? The "page" fetching this is our extension.

lvyitian replied:
Oh, then that won't be a problem.

lvyitian replied:
But what if the website only supports http and not https.

Korne127 replied:
I thought about that for a longer time too. But Ruffle is used to play Flash games, mostly legacy content. Old SWF files are especially likely to be hosted on outdated websites which don't support HTTPS yet.
And if the website uses HTTPS, that's just used. HTTP is only used if it's the websites only supports it, but not HTTPS.

Korne127 committed:
web: Fix & Improve extension player URL error messages

Previously, the Ruffle player had shown an incorrect error message if
the entered URL wasn't a valid web URL.
This has been fixed and improved. Three new error messages have been
added for this case: One in case the entered URL is a local file URL,
one in case the entered URL uses a different protocol than https, http
or file and one in case the entered URL isn't a valid URL for any
protocol at all.
Additionally, the error message in case Ruffle couldn't load the SWF
file (using a valid web URL) has been improved to also include the tip
that the HTTP protocol should be (explicitly) used for websites that
don't support HTTPS.

Korne127 committed:
web: Display loading screen while resolving the SWF URL

Previously, when entering a URL in the extension player, it had been
resolved before the loading screen had been displayed. As resolving can
include testing for server availability (to decide whether HTTPS or HTTP
is used), this could create a wait time without the loading screen.
This has been fixed. The loading screen is now displayed while the URL
is resolved. Additionally, the timeouts have been increased (they were
low because of the lack of a loading animation).

A bug that caused the document title to not be set correctly if the SWF
URL was the main page of a domain has also been fixed.

Korne127 committed:
web: Display resolved URL instead of input URL

After entering a URL in the extension player, Ruffle resolves it,
potentially adding a protocol. This resolved URL is now displayed in the
URL input element instead of the original input URL.
This improves the transparency of which protocol is used and makes it
easier for the user to change it.

Korne127 commented:
I rebased the Pull Request and adapted it accordingly to @kjarosh's suggestions.

kjarosh commented on web/packages/extension/assets/css/common.css:
```diff
@@ -79,6 +79,10 @@ button {
 }
 
+button:hover {
+    background: var(--ruffle-blue);
```
That unfortunately doesn't look good in some places, because the hovered button has the same color as the background and it disappears, e.g.

![image](https://github.com/user-attachments/assets/9487abaa-70ef-48a8-a48b-0022fe6f0656)

I would suggest the following:

```suggestion
    filter: brightness(1.2);
```

![image](https://github.com/user-attachments/assets/0527fc9b-b1b9-4a9d-98ea-ea76264ba591)

Korne127 replied:
Thanks for the notice, I actually just intended this for the modal buttons.
I updated it to account for these buttons as well.

Korne127 commented on web/packages/core/src/internal/ui/panic.tsx:
```diff
@@ -154,14 +155,29 @@ function createPanicError(error: Error | null): {
+                    url: error.swfUrl.href.replace(urlExtensionPart, ""),
+                }),
+                actions: [CommonActions.ShowDetails],
+            };
```
Thanks for noticing this! I fixed it so that all these extension player - specific error messages (as well as the http advice) are only shown in the extension player and a new additional error message is displayed if a website is misconfigured and uses a different protocol.

Korne127 committed:
web: Don't try to connect to an unsupported URL

If Ruffle recognises that a URL is invalid or not supported, it now
doesn't try to connect to it, but instantly panics.
This reduces unnecessary connection attempts and improves reaction
speed.

Korne127 committed:
web: Fix error messages outside the extension player

The error messages outside the extension player have been fixed.
Ruffle now correctly determines whether the last load command originated
from the extension player and displays the error messages specific to it
(invalid URL / local URL / URL with unsupported protocol entered) only
in that case. The advice to try entering http explicitly has also been
removed if the URL is not loaded from the extension player.

Additionally, if the website now embeds a URL with an unsupported
protocol (e.g. file), a new specific error message about the website
being misconfigured is now displayed instead of a button to open the URL
in the extension player (which itself will result in an error).

Korne127 committed:
web: Add OriginAPI to the web player

The new OriginAPI has been added to the web player.
It is used to access behaviour that's specific to the Ruffle origin
(selfhosted, demo or extension).
Unlike the SourceAPI, which manages the whole Ruffle polyfill and
player, it provides minor helper methods and is given to the (internal)
Ruffle player.

The OriginAPI has three implementations: SelfhostedOrigin, DemoOrigin
and ExtensionOrigin. The respective instance is created when installing
the Ruffle instance. When using SourceAPI#createPlayer or #polyfill, the
OriginAPI is given to the then-created (internal) player. This is done
by calling the new RufflePlayerElement#initialize method after creating
it.
The player API now also provides a method to access its OriginAPI.

The previous extension related variables and functions (like
inExtensionPlayer) have been removed from the core code. Instead,
createPanicError now uses OriginAPI calls to display the correct error
message.

Additionally, internalSourceApi and its usages have been replaced by
several functions, reducing the likelihood of future mistakes.

The documentation and code quality have also been generally improved.

Korne127 commented:
Extension-player specific error messages (as well as the http advice) are now only shown in the extension player, and a new additional error message is displayed if a website is misconfigured and uses a different protocol.

This is done by creating a so-called `OriginAPI` which allows the web core to access behaviour that's specific to the Ruffle origin (selfhosted, demo or extension).
For more information, see the last commit message.
